File: backend/biz/schema/expert_indirect_relation.py
import re
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

IndirectRelationType = Literal["学术关联", "机构关联", "项目关联"]


class ExpertIndirectRelationRequest(BaseModel):
    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "core_node_id": "4G7t0B0t",
                "relation_types": "学术关联",
                "path_depth": 2,
                "min_strength": 0.65,
            }
        }
    )

    core_node_id: str = Field(
        ...,
        min_length=1,
        max_length=64,
        description="核心专家或人才节点 ID。",
    )
    relation_types: list[IndirectRelationType] = Field(
        ...,
        min_length=1,
        max_length=1,
        description="需要保留的间接关系类型，必须且只能选择：学术关联、机构关联或项目关联中的一项。",
    )
    path_depth: int = Field(default=2, ge=2, le=3, description="路径分析深度，支持 2-3 跳。")
    min_strength: float = Field(
        default=0.65,
        ge=0,
        le=1,
        description="最小关联强度阈值。",
    )

    @field_validator("core_node_id", mode="before")
    @classmethod
    def normalize_core_node_id(cls, value: str) -> str:
        if value is None:
            return value
        value = str(value)
        if re.search(r"\s", value):
            raise ValueError("核心节点 ID 不能包含空格或 !@#￥%& 等异常字符")
        if len(value) > 64:
            raise ValueError("核心节点 ID 长度不能超过 64 个字符")
        if not re.fullmatch(r"[\w\u4e00-\u9fff·.\-]+", value):
            raise ValueError("核心节点 ID 不能包含空格或 !@#￥%& 等异常字符")
        return value

    @field_validator("relation_types", mode="before")
    @classmethod
    def normalize_relation_types(cls, value: Any) -> list[str]:
        if value in (None, ""):
            return []
        if isinstance(value, str):
            normalized = value.replace("；", ";").replace("，", ",").replace("、", ",")
            return [
                item.strip() for item in normalized.replace(";", ",").split(",") if item.strip()
            ]
        return value


class IndirectNode(BaseModel):
    id: str
    name: str
    entityType: str
    labels: list[str] = Field(default_factory=list)
    properties: dict[str, Any] = Field(default_factory=dict)


class IndirectEdge(BaseModel):
    id: str
    type: str
    source: str
    target: str
    properties: dict[str, Any] = Field(default_factory=dict)


class IndirectRelationPath(BaseModel):
    pathId: str
    depth: int
    relationType: str
    strength: float
    pathText: str
    targetNode: IndirectNode
    nodes: list[IndirectNode]
    edges: list[IndirectEdge]


class StructuredIndirectRelationResult(BaseModel):
    coreNode: IndirectNode
    pathDepth: int
    defaultPathDepth: int = Field(
        default=2,
        description="系统默认路径深度；请求未填写 path_depth 时采用 2 跳。",
    )
    minStrength: float
    directNodeCount: int
    indirectNodeCount: int
    pathCount: int
    relationTypeCount: dict[str, int]
    averageStrength: float
    maxStrength: float
    directNodes: list[IndirectNode]
    indirectNodes: list[IndirectNode]
    paths: list[IndirectRelationPath]


class IndirectProvenanceEvidence(BaseModel):
    title: str
    sourceTable: str
    sourceField: str
    graphVid: str


class IndirectProvenance(BaseModel):
    sourceDatabase: str
    summary: str
    evidences: list[IndirectProvenanceEvidence] = Field(default_factory=list)


class ExpertIndirectRelationResponse(BaseModel):
    structuredResult: StructuredIndirectRelationResult
    provenance: IndirectProvenance
    rules: list[dict[str, Any]] = Field(default_factory=list)

File: frontend/src/api/expertIndirectRelation.ts
import { http } from './http'

export interface IndirectProvenanceEvidence {
  title: string
  sourceTable: string
  sourceField: string
  graphVid: string
  businessTable?: string
  technicalTable?: string
  recordId?: string
  fieldIdentifier?: string
  summary?: string
}

export interface IndirectProvenance {
  sourceDatabase: string
  summary: string
  evidences: IndirectProvenanceEvidence[]
}

export interface IndirectNode {
  id: string
  name: string
  entityType: string
  labels: string[]
  properties: Record<string, unknown>
}

export interface IndirectEdge {
  id: string
  type: string
  source: string
  target: string
  properties: Record<string, unknown>
}

export interface IndirectRelationPath {
  pathId: string
  depth: number
  relationType: string
  strength: number
  pathText: string
  targetNode: IndirectNode
  nodes: IndirectNode[]
  edges: IndirectEdge[]
}

export interface ExpertIndirectRelationResult {
  coreNode: IndirectNode
  pathDepth: number
  defaultPathDepth: number
  minStrength: number
  directNodeCount: number
  indirectNodeCount: number
  pathCount: number
  relationTypeCount: Record<string, number>
  averageStrength: number
  maxStrength: number
  directNodes: IndirectNode[]
  indirectNodes: IndirectNode[]
  paths: IndirectRelationPath[]
}

export interface ExpertIndirectRelationResponse {
  structuredResult: ExpertIndirectRelationResult
  provenance: IndirectProvenance
  rules: Array<Record<string, any>>
}

export interface ExpertIndirectRelationRequest {
  core_node_id: string
  relation_types: string[]
  path_depth: number
  min_strength: number
}

const ENDPOINT = '/v1/kg-construction/expert-indirect-relations/demo/structured-result'
const ANNOTATION_ENDPOINT = '/v1/kg-construction/expert-indirect-relations/annotations'

export const analyzeExpertIndirectRelation = (
  payload: ExpertIndirectRelationRequest,
) => http.post<ExpertIndirectRelationResponse>(ENDPOINT, payload) as unknown as Promise<ExpertIndirectRelationResponse>

export interface IndirectRelationAnnotationItem {
  sourceVid: string
  targetVid: string
  annotation: string
  updateTime?: string | null
}

interface AnnotationApiEnvelope {
  code: number
  success: boolean
  msg?: string
}

/** 边主键：两端 VID 按字典序归一后拼接，与后端 kg_indirect_relation_annotation 复合主键一致。 */
export const indirectRelationEdgeKey = (vidA: string, vidB: string): string => {
  const [source, target] = [vidA, vidB].sort()
  return `${source}:${target}`
}

function ensureAnnotationSuccess(resp: AnnotationApiEnvelope | undefined): void {
  if (!resp || resp.success === false || (resp.code !== undefined && resp.code !== 200)) {
    throw new Error(resp?.msg || '关系标注接口返回失败')
  }
}

/** 批量查询边主键对应的关系标注；查不到的边即无标注（不出现在返回 map 中，展示为空）。 */
export async function fetchIndirectRelationAnnotations(
  edgeKeys: string[],
): Promise<Record<string, string>> {
  if (!edgeKeys.length) return {}
  const resp = (await http.get(ANNOTATION_ENDPOINT, {
    params: { edges: edgeKeys.join(',') },
  })) as unknown as AnnotationApiEnvelope & {
    data?: { items?: IndirectRelationAnnotationItem[] }
  }
  ensureAnnotationSuccess(resp)
  const annotations: Record<string, string> = {}
  for (const item of resp.data?.items ?? []) {
    annotations[indirectRelationEdgeKey(item.sourceVid, item.targetVid)] =
      item.annotation || ''
  }
  return annotations
}

/** 保存（upsert）一条关系标注到业务库；annotation 传空串表示清除标注。 */
export async function upsertIndirectRelationAnnotation(
  sourceVid: string,
  targetVid: string,
  annotation: string,
): Promise<IndirectRelationAnnotationItem> {
  const resp = (await http.post(ANNOTATION_ENDPOINT, {
    sourceVid,
    targetVid,
    annotation,
  })) as unknown as AnnotationApiEnvelope & {
    data: IndirectRelationAnnotationItem
  }
  ensureAnnotationSuccess(resp)
  return resp.data
}
